text/0000-enum-variant-types.md

```diff
@@ -0,0 +1,203 @@
+be at odds with expectations. However, since they generally behave identically, this should not
+prove to be a significant problem.
+- As variant types need to include generic parameter information that is not necessarily included in
+their definitions, it will be necessary to include explicit type annotations more often than is
+typical. Although this is unfortunate, it is necessary to preserve all the desirable behaviour of
+variant types described here: namely complete backwards-compatibility precise type inference
+(e.g. allowing `x` in `let x = Sum::A;` to have type `Sum::A` without explicit type annotations).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The advantages of this approach are:
+- It naturally allows variants to be treated as types, intuitively.
+- It doesn't require explicit type annotations to reap the benefits of variant types.
+- As variant types and enum types are represented identically, there are no coercion costs.
+- It doesn't require type fallback.
+- It doesn't require value-tracking or complex type system additions such as refinement types.
+
+One obvious alternative is to represent variant types differently to enum types and then coerce them
+when used as an enum. This could potentially reduce memory overhead for smaller variants
+(additionally no longer requiring the discriminant to be stored) and reduce the issue with providing
+irrelevant type parameters. However, it makes coercion more expensive and complex (as a variant
+could coerce to various enum types depending on the unspecified generic parameters).
+
+Variant types have [previously been proposed for Rust](https://github.com/rust-lang/rfcs/pull/1450).
+However, it used a more complex type inference procedure based on fallback and permitted fallible
+coercion to variant types. The method proposed here is implementationally simpler and more
+intuitive.
+
+# Prior art
+[prior-art]: #prior-art
+
+Type-theoretically, enums are sum types. A sum type `S := A + B` is a type, `S`, defined in relation
+to two other types, `A` and `B`. Variants are specifically types, but in programming it's usually
+useful to consider particular variants in relation to each other, rather than standalone (which is
+why `enum` *defines* types for its variants rather than using pre-existing types for its variants).
+
+However, it is often useful to briefly consider these variant types alone, which is what this
+RFC proposes.
+
+Although sum types are becoming increasingly common in programming languages, most do not choose to
+allow the variants to be treated as types in their own right. However, we propose that the patterns
+in Rust make variant types more appealing than they might be in other programming languages with
+variant types.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- None.
```

Comment:
Any [Future possibilities](https://github.com/rust-lang/rfcs/pull/2561)?

Reply:
Oh, I split this off before there was such a section. I'll add one.